src/full/Agda/TypeChecking/Unquote.hs

```diff
@@ -1041,6 +1044,35 @@ evalTCM v = do
       locallyReduceAllDefs $ checkFunDef NotDelayed i x cs
       primUnitUnit
 
+    tcPragmaForeign :: Text -> Text -> TCM Term
+    tcPragmaForeign backend code = do
+      modifyTC $ \ st -> let preScSt = stPreScopeState st in st
+        { stPreScopeState = preScSt
+          { stPreForeignCode = Map.insertWith
+            (flip (++))
+            (T.unpack backend)
+            [ForeignCode noRange $ T.unpack code]
+            (stPreForeignCode preScSt)
+          }
+        }
```

Comment:
Looks like you are doing _lens_ stuff.  Maybe factor our a lens here (if you cannot find a suitable one already).